docs/docs/typelevel.md

```diff
@@ -0,0 +1,449 @@
+
+One might criticize that this scheme involves code duplication. In the example above, the recursive `concat` algorithm had to be implemented twice, once as a regular function, the other time as a transparent function. However, in practice it is is quite likely that the regular function would use optimized data representatons and algortihms that do not lend themselves easily to a typelevel interpretation. In these cases a dual implementation is required anyway.
+
+## Code Specialization
+
+Transparent functions are a convenient means to achieve code specialization. As an example, consider implementing a math library that implements (among others) a `dotProduct` method. Here is a possible implementation of `MathLib` with some user code.
+```scala
+abstract class MathLib[N : Numeric] {
+  def dotProduct(xs: Array[N], ys: Array[N]): N
+}
+object MathLib {
+
+  transparent def apply[N](implicit n: Numeric[N]) = new MathLib[N] {
+    import n._
+    def dotProduct(xs: Array[N], ys: Array[N]): N = {
+      require(xs.length == ys.length)
+      var i = 0
+      var s: N = n.zero
+      while (i < xs.length) {
+        s = s + xs(i) * ys(i)
+        i += 1
+      }
+      s
+    }
+  }
+}
+```
+`MathLib` is intentionally kept very abstract - it works for any element type `N` for which a `math.Numeric` implementation exists.
+Here is some code that uses `MathLib`:
+```scala
+val mlib = MathLib[Double]
+
+val xs = Array(1.0, 1.0)
+val ys = Array(2.0, -3.0)
+mlib.dotProduct(xs, ys)
+```
+The implementation code for a given numeric type `N` is produced by the `apply` method of `MathLib`.
+Even though the `dotProduct` code looks simple, there is a lot of hidden complexity in the generic code:
+
+ - It uses unbounded generic arrays, which means code on the JVM needs reflection to access their elements
+ - All operations on array elements forward to generic operations in class `Numeric`.
+
+It would be quite hard for even a good optimizer to undo the generic abstractions and replace them with something simpler if `N` is specialized to a primitive type like `Double`. But since `apply` is marked `transparent`, the specialization happens automatically as a result of inlining the body of `apply` with the new types. Indeed, the specialized version of `dotProduct` looks like this:
+```
+def dotProduct(xs: Array[Double], ys: Array[Double]): Double = {
+  require(xs.length == ys.length)
+  var i = 0
+  var s: Double = n.zero
+  while (i < xs.length) {
+    s = s + xs(i) * ys(i)
+    i += 1
+  }
+  s
+}
+```
+In other words, specialization with transparent functions allows "abstraction without regret". The price to pay for this
+is the increase of code size through inlining. That price is often worth paying, but inlining decisions need to be considered carefully.
+
```

Comment:
I see how this can look like a lack of hygiene, but it isn't. Hygiene is about free variables being rebound, but `*` is not a free variable: it is a method of its left operand, that, in general depends on the type of the operand. Arguably, it would be a surprise if `*` was _not_ rebound. I still have to write this argument up in detail that in general, but maybe it's better to get more experience with the system first.